compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -376,19 +397,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expected: Expectation<'tcx>,
         expr: &'tcx hir::Expr<'tcx>,
     ) -> Ty<'tcx> {
-        let hint = expected.only_has_type(self).map_or(NoExpectation, |ty| {
+        let type_ascr_ctxt = expected.get_coercion_ctxt();
+        let hint = expected.only_has_type(self).map_or(NoExpectation(type_ascr_ctxt), |ty| {
             match ty.kind() {
                 ty::Ref(_, ty, _) | ty::RawPtr(ty::TypeAndMut { ty, .. }) => {
                     if oprnd.is_syntactic_place_expr() {
                         // Places may legitimately have unsized types.
                         // For example, dereferences of a fat pointer and
                         // the last field of a struct can be unsized.
-                        ExpectHasType(ty)
+                        ExpectHasType(ty, type_ascr_ctxt)
                     } else {
-                        Expectation::rvalue_hint(self, ty)
+                        Expectation::rvalue_hint(self, ty, type_ascr_ctxt)
                     }
                 }
-                _ => NoExpectation,
+                _ => NoExpectation(type_ascr_ctxt),
```

Comment:
This should absolutely *not* be a coercion site. I think this suggests a missing test in the PR. Something like the following:

```rust
fn foo<'a>(a: &'a u32) -> &'static u32 {
    let mut x: &'static u32 = &22;
    let y: &mut &'a u32 = &mut (x: &'a u32);
    *y = a;
    x
}
```

Does that compile with your PR? If so, I think it would be fixed by changing `type_ascr_ctxt` to not be considered a coercion site.